src/jitlayers.cpp

```diff
@@ -764,6 +762,24 @@ static void jl_emit_codeinst_to_jit(
     jl_unique_gcsafe_lock lock(engine_lock);
     if (invokenames.count(codeinst) || jl_is_compiled_codeinst(codeinst))
         return; // destroy everything
+    const std::string &specf = decls.specFunctionObject;
+    const std::string &f = decls.functionObject;
+    assert(!f.empty());
+    // Prepare debug info to receive this function
+    // record that this function name came from this linfo,
+    // so we can build a reverse mapping for debug-info.
+    bool toplevel = !jl_is_method(jl_get_ci_mi(codeinst)->def.method);
+    if (!toplevel) {
+        // don't remember toplevel thunks because
+        // they may not be rooted in the gc for the life of the program,
```

Comment:
Should we use a finalizer for this instead? (thread safety might make this annoying) Though it would be quite a bit of added complexity

Reply:
We could, but that introduces some other consistency issues too since the lookup into this map is async. It is probably best just to keep it the way things were.